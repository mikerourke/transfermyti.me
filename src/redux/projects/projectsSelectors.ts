import { isNil, pluck, propOr } from "ramda";

import { selectIdToLinkedId } from "~/api/selectIdToLinkedId";
import {
  areExistsInTargetShownSelector,
  toolNameByMappingSelector,
} from "~/redux/allEntities/allEntitiesSelectors";
import { createSelector } from "~/redux/reduxTools";
import { sourceTimeEntryCountByIdFieldSelectorFactory } from "~/redux/timeEntries/timeEntriesSelectors";
import { activeWorkspaceIdSelector } from "~/redux/workspaces/workspacesSelectors";
import type { EntityTableRecord, Project, ReduxState } from "~/typeDefs";

export const sourceProjectsByIdSelector = createSelector(
  (state: ReduxState) => state.projects.source,
  (sourceProjectsById): Dictionary<Project> => sourceProjectsById,
);

export const sourceProjectsSelector = createSelector(
  sourceProjectsByIdSelector,
  (sourceProjectsById): Project[] => Object.values(sourceProjectsById),
);

export const targetProjectsByIdSelector = createSelector(
  (state: ReduxState) => state.projects.target,
  (targetProjectsById): Dictionary<Project> => targetProjectsById,
);

const targetProjectsSelector = createSelector(
  targetProjectsByIdSelector,
  (targetProjectsById): Project[] => Object.values(targetProjectsById),
);

export const includedSourceProjectsSelector = createSelector(
  sourceProjectsSelector,
  (sourceProjects): Project[] =>
    sourceProjects.filter((sourceProject) => sourceProject.isIncluded),
);

export const includedSourceProjectIdsSelector = createSelector(
  includedSourceProjectsSelector,
  (sourceProjects): string[] => pluck("id", sourceProjects),
);

export const sourceProjectsForTransferSelector = createSelector(
  includedSourceProjectsSelector,
  (sourceProjects): Project[] =>
    sourceProjects.filter((sourceProject) => isNil(sourceProject.linkedId)),
);

export const sourceProjectsInActiveWorkspaceSelector = createSelector(
  sourceProjectsSelector,
  activeWorkspaceIdSelector,
  (sourceProjects, activeWorkspaceId) =>
    sourceProjects.filter(
      (sourceProject) => sourceProject.workspaceId === activeWorkspaceId,
    ),
);

export const projectIdToLinkedIdSelector = createSelector(
  sourceProjectsByIdSelector,
  (sourceProjectsById): Dictionary<string> =>
    selectIdToLinkedId(sourceProjectsById),
);

export const projectsForInclusionsTableSelector = createSelector(
  areExistsInTargetShownSelector,
  sourceProjectsInActiveWorkspaceSelector,
  targetProjectsByIdSelector,
  sourceTimeEntryCountByIdFieldSelectorFactory("projectId"),
  (
    areExistsInTargetShown,
    sourceProjects,
    targetProjectsById,
    timeEntryCountByProjectId,
  ): EntityTableRecord<Project>[] => {
    const projectTableRecords: EntityTableRecord<Project>[] = [];

    for (const sourceProject of sourceProjects) {
      const existsInTarget = sourceProject.linkedId !== null;

      if (existsInTarget && !areExistsInTargetShown) {
        continue;
      }

      let isActiveInTarget = false;
      if (existsInTarget) {
        const targetId = sourceProject.linkedId as string;

        isActiveInTarget = targetProjectsById[targetId].isActive;
      }

      const entryCount = propOr<number, Dictionary<number>, number>(
        0,
        sourceProject.id,
        timeEntryCountByProjectId,
      );

      projectTableRecords.push({
        ...sourceProject,
        entryCount,
        existsInTarget,
        isActiveInSource: sourceProject.isActive,
        isActiveInTarget,
      });
    }

    return projectTableRecords;
  },
);

export const projectsTotalCountsByTypeSelector = createSelector(
  projectsForInclusionsTableSelector,
  (projectsForInclusionsTable) =>
    projectsForInclusionsTable.reduce(
      (
        acc,
        {
          entryCount,
          existsInTarget,
          isIncluded,
          isActiveInSource,
          isActiveInTarget,
        }: EntityTableRecord<Project>,
      ) => ({
        entryCount: acc.entryCount + entryCount,
        existsInTarget: acc.existsInTarget + (existsInTarget ? 1 : 0),
        isActiveInSource: acc.isActiveInSource + (isActiveInSource ? 1 : 0),
        isActiveInTarget: acc.isActiveInTarget + (isActiveInTarget ? 1 : 0),
        isIncluded: acc.isIncluded + (isIncluded ? 1 : 0),
      }),
      {
        entryCount: 0,
        existsInTarget: 0,
        isActiveInSource: 0,
        isActiveInTarget: 0,
        isIncluded: 0,
      },
    ),
);

export const projectsByWorkspaceIdByToolNameSelector = createSelector(
  toolNameByMappingSelector,
  sourceProjectsSelector,
  targetProjectsSelector,
  (toolNameByMapping, sourceProjects, targetProjects) => ({
    [toolNameByMapping.source]: groupProjectsByWorkspaceId(sourceProjects),
    [toolNameByMapping.target]: groupProjectsByWorkspaceId(targetProjects),
  }),
);

function groupProjectsByWorkspaceId(
  projects: Project[],
): Dictionary<Project[]> {
  const projectsByWorkspaceId: Dictionary<Project[]> = {};

  for (const project of projects) {
    const workspaceProjects = propOr<
      Project[],
      Dictionary<Project[]>,
      Project[]
    >([], project.workspaceId, projectsByWorkspaceId);

    projectsByWorkspaceId[project.workspaceId] = [
      ...workspaceProjects,
      project,
    ];
  }

  return projectsByWorkspaceId;
}
