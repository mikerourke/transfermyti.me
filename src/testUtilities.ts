export { FAKES } from "~/redux/__fakes__/state";

type MockInstance = jest.MockInstance<AnyValid, AnyValid>;

interface WrapForMockingReturn {
  calls: AnyValid[];
  clear(): MockInstance;
  implementOnce(funcOrReturn: AnyValid): MockInstance;
  rejectAlways(err?: AnyValid): MockInstance;
  rejectOnce(err?: AnyValid): MockInstance;
  resolveMultiple(values: AnyValid[]): void;
  resolveOnce(value?: AnyValid): MockInstance;
  returnOnce(value?: AnyValid): MockInstance;
  returnMultiple(values: AnyValid[]): void;
  throwOnce(message?: string): MockInstance;
}

/**
 * Wrappers around someFunc.mock...() that Jest provides. Using
 * this avoids the annoying `someFunc as AnyValid` required to shut TypeScript
 * up.
 */
export function wrapForMocking(func: AnyValid): WrapForMockingReturn {
  return {
    calls: func?.mock?.calls,
    clear(): MockInstance {
      return func.mockClear();
    },
    implementOnce(funcOrReturn: AnyValid): MockInstance {
      if (typeof funcOrReturn === "function") {
        return func.mockImplementationOnce(funcOrReturn);
      } else {
        return func.mockImplementationOnce(() => funcOrReturn);
      }
    },
    rejectAlways(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValue(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    rejectOnce(err: AnyValid = "error"): MockInstance {
      return func.mockRejectedValueOnce(
        typeof err === "string" ? new Error(err) : err,
      );
    },
    resolveMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockResolvedValueOnce(value);
      }
    },
    resolveOnce(value: AnyValid = null): MockInstance {
      return func.mockResolvedValueOnce(value);
    },
    returnOnce(value: AnyValid = null): MockInstance {
      return func.mockReturnValueOnce(value);
    },
    returnMultiple(values: AnyValid[]): void {
      for (const value of values) {
        func.mockReturnValueOnce(value);
      }
    },
    throwOnce(message: string = "error"): MockInstance {
      return func.mockImplementationOnce(() => {
        throw new Error(message);
      });
    },
  };
}

/**
 * Allows you to get around `delay()` calls in sagas.
 * @see https://github.com/jfairbank/redux-saga-test-plan/issues/257#issuecomment-503016758
 * @example
 *   return expectSaga(someSaga)
 *     .provide([
 *       { call: provideDelay },
 *     ])
 *     ...
 */
export function provideDelay(
  { fn }: AnyValid,
  next: AnyValid,
): null | VoidFunction {
  return fn.name === "delayP" ? null : next();
}
