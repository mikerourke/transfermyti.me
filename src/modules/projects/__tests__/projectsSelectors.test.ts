import cases from "jest-in-case";

import { FAKES } from "~/testUtilities";

import * as projectsSelectors from "../projectsSelectors";

const MOCK_STATE = { ...FAKES.REDUX_STATE };

describe("within projectsSelectors", () => {
  cases(
    "selectors that directly access state return the correct value",
    (options) => {
      const result = options.selector(MOCK_STATE);

      expect(result).toEqual(options.expected);
    },
    [
      {
        name: "sourceProjectsByIdSelector returns state.sourceProjectsById",
        selector: projectsSelectors.sourceProjectsByIdSelector,
        expected: MOCK_STATE.projects.source,
      },
      {
        name: "sourceProjectsSelector returns values from state.sourceProjectsById",
        selector: projectsSelectors.sourceProjectsSelector,
        expected: Object.values(MOCK_STATE.projects.source),
      },
      {
        name: "targetProjectsByIdSelector returns state.targetProjectsById",
        selector: projectsSelectors.targetProjectsByIdSelector,
        expected: MOCK_STATE.projects.target,
      },
    ],
  );

  cases(
    "the selectors match their snapshots",
    (options) => {
      const result = options.selector(MOCK_STATE);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "for the includedSourceProjectsSelector",
        selector: projectsSelectors.includedSourceProjectsSelector,
      },
      {
        name: "for the includedSourceProjectIdsSelector",
        selector: projectsSelectors.includedSourceProjectIdsSelector,
      },
      {
        name: "for the sourceProjectsForTransferSelector",
        selector: projectsSelectors.sourceProjectsForTransferSelector,
      },
      {
        name: "for the sourceProjectsInActiveWorkspaceSelector",
        selector: projectsSelectors.sourceProjectsInActiveWorkspaceSelector,
      },
      {
        name: "for the projectIdToLinkedIdSelector",
        selector: projectsSelectors.projectIdToLinkedIdSelector,
      },
      {
        name: "for the projectsTotalCountsByTypeSelector",
        selector: projectsSelectors.projectsTotalCountsByTypeSelector,
      },
      {
        name: "for the projectsByWorkspaceIdByToolNameSelector",
        selector: projectsSelectors.projectsByWorkspaceIdByToolNameSelector,
      },
    ],
  );

  cases(
    "the projectsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown",
    (options) => {
      const updatedState = {
        ...MOCK_STATE,
        allEntities: {
          ...MOCK_STATE.allEntities,
          areExistsInTargetShown: options.areExistsInTargetShown,
        },
      };
      const result = projectsSelectors.projectsForInclusionsTableSelector(updatedState);

      expect(result).toMatchSnapshot();
    },
    [
      {
        name: "when state.allEntities.areExistsInTargetShown = true",
        areExistsInTargetShown: true,
      },
      {
        name: "when state.allEntities.areExistsInTargetShown = false",
        areExistsInTargetShown: false,
      },
    ],
  );
});
