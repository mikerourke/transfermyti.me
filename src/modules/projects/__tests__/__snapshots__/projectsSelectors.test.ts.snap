// Vitest Snapshot v1

exports[`within projectsSelectors > the projectsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown > when state.allEntities.areExistsInTargetShown = false 1`] = `
[
  {
    "clientId": "3001",
    "color": "#ea468d",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": true,
    "id": "2001",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-01",
    "memberOf": "projects",
    "name": "Test Project A",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#3750b5",
    "entryCount": 3,
    "estimate": {
      "estimate": 40,
      "type": "MANUAL",
    },
    "existsInTarget": true,
    "id": "2002",
    "isActive": false,
    "isActiveInSource": false,
    "isActiveInTarget": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-02",
    "memberOf": "projects",
    "name": "Test Project B",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#c56bff",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": true,
    "id": "2003",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-03",
    "memberOf": "projects",
    "name": "Test Project C",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2004",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2005",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#a01aa5",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2006",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project F",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors > the projectsForInclusionsTableSelector matches its snapshot based on state.allEntities.areExistsInTargetShown > when state.allEntities.areExistsInTargetShown = false 2`] = `
[
  {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2004",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2005",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#a01aa5",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "existsInTarget": false,
    "id": "2006",
    "isActive": true,
    "isActiveInSource": true,
    "isActiveInTarget": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project F",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors > the selectors match their snapshots > for the projectIdToLinkedIdSelector 1`] = `
[
  {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2004",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2005",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors > the selectors match their snapshots > for the projectIdToLinkedIdSelector 2`] = `
[
  "2004",
  "2005",
]
`;

exports[`within projectsSelectors > the selectors match their snapshots > for the projectIdToLinkedIdSelector 3`] = `
[
  {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2004",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2005",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors > the selectors match their snapshots > for the projectIdToLinkedIdSelector 4`] = `
[
  {
    "clientId": "3001",
    "color": "#ea468d",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2001",
    "isActive": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-01",
    "memberOf": "projects",
    "name": "Test Project A",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#3750b5",
    "entryCount": 3,
    "estimate": {
      "estimate": 40,
      "type": "MANUAL",
    },
    "id": "2002",
    "isActive": false,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-02",
    "memberOf": "projects",
    "name": "Test Project B",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#c56bff",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2003",
    "isActive": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": "clock-project-03",
    "memberOf": "projects",
    "name": "Test Project C",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#04bb9b",
    "entryCount": 3,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2004",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project D",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#205500",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2005",
    "isActive": true,
    "isBillable": false,
    "isIncluded": true,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project E",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
  {
    "clientId": null,
    "color": "#a01aa5",
    "entryCount": 4,
    "estimate": {
      "estimate": 0,
      "type": "MANUAL",
    },
    "id": "2006",
    "isActive": true,
    "isBillable": false,
    "isIncluded": false,
    "isPublic": false,
    "linkedId": null,
    "memberOf": "projects",
    "name": "Test Project F",
    "userIds": [
      "6001",
    ],
    "workspaceId": "1001",
  },
]
`;

exports[`within projectsSelectors > the selectors match their snapshots > for the projectIdToLinkedIdSelector 5`] = `
{
  "2001": "clock-project-01",
  "2002": "clock-project-02",
  "2003": "clock-project-03",
  "clock-project-01": "2001",
  "clock-project-02": "2002",
  "clock-project-03": "2003",
}
`;

exports[`within projectsSelectors > the selectors match their snapshots > for the projectsByWorkspaceIdByToolNameSelector 1`] = `
{
  "entryCount": 20,
  "existsInTarget": 3,
  "isActiveInSource": 5,
  "isActiveInTarget": 2,
  "isIncluded": 2,
}
`;

exports[`within projectsSelectors > the selectors match their snapshots > for the projectsByWorkspaceIdByToolNameSelector 2`] = `
{
  "clockify": {
    "clock-workspace-01": [
      {
        "clientId": "clock-client-01",
        "color": "#ea468d",
        "entryCount": 0,
        "estimate": {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-01",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2001",
        "memberOf": "projects",
        "name": "Test Project A",
        "userIds": [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
      {
        "clientId": "clock-client-02",
        "color": "#c56bff",
        "entryCount": 0,
        "estimate": {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-02",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2002",
        "memberOf": "projects",
        "name": "Test Project B",
        "userIds": [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
      {
        "clientId": "",
        "color": "#04bb9b",
        "entryCount": 0,
        "estimate": {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-03",
        "isActive": false,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2003",
        "memberOf": "projects",
        "name": "Test Project C",
        "userIds": [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
      {
        "clientId": "",
        "color": "#04bb9b",
        "entryCount": 0,
        "estimate": {
          "estimate": "PT0S",
          "type": "MANUAL",
        },
        "id": "clock-project-04",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2004",
        "memberOf": "projects",
        "name": "Test Project D",
        "userIds": [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
      {
        "clientId": "",
        "color": "#205500",
        "entryCount": 0,
        "estimate": {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-05",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2005",
        "memberOf": "projects",
        "name": "Test Project E",
        "userIds": [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
      {
        "clientId": "",
        "color": "#a01aa5",
        "entryCount": 0,
        "estimate": {
          "estimate": "PT0S",
          "type": "AUTO",
        },
        "id": "clock-project-06",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "2006",
        "memberOf": "projects",
        "name": "Test Project F",
        "userIds": [
          "clock-user-01",
        ],
        "workspaceId": "clock-workspace-01",
      },
    ],
  },
  "toggl": {
    "1001": [
      {
        "clientId": "3001",
        "color": "#ea468d",
        "entryCount": 3,
        "estimate": {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2001",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "clock-project-01",
        "memberOf": "projects",
        "name": "Test Project A",
        "userIds": [
          "6001",
        ],
        "workspaceId": "1001",
      },
      {
        "clientId": null,
        "color": "#3750b5",
        "entryCount": 3,
        "estimate": {
          "estimate": 40,
          "type": "MANUAL",
        },
        "id": "2002",
        "isActive": false,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "clock-project-02",
        "memberOf": "projects",
        "name": "Test Project B",
        "userIds": [
          "6001",
        ],
        "workspaceId": "1001",
      },
      {
        "clientId": null,
        "color": "#c56bff",
        "entryCount": 3,
        "estimate": {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2003",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": "clock-project-03",
        "memberOf": "projects",
        "name": "Test Project C",
        "userIds": [
          "6001",
        ],
        "workspaceId": "1001",
      },
      {
        "clientId": null,
        "color": "#04bb9b",
        "entryCount": 3,
        "estimate": {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2004",
        "isActive": true,
        "isBillable": false,
        "isIncluded": true,
        "isPublic": false,
        "linkedId": null,
        "memberOf": "projects",
        "name": "Test Project D",
        "userIds": [
          "6001",
        ],
        "workspaceId": "1001",
      },
      {
        "clientId": null,
        "color": "#205500",
        "entryCount": 4,
        "estimate": {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2005",
        "isActive": true,
        "isBillable": false,
        "isIncluded": true,
        "isPublic": false,
        "linkedId": null,
        "memberOf": "projects",
        "name": "Test Project E",
        "userIds": [
          "6001",
        ],
        "workspaceId": "1001",
      },
      {
        "clientId": null,
        "color": "#a01aa5",
        "entryCount": 4,
        "estimate": {
          "estimate": 0,
          "type": "MANUAL",
        },
        "id": "2006",
        "isActive": true,
        "isBillable": false,
        "isIncluded": false,
        "isPublic": false,
        "linkedId": null,
        "memberOf": "projects",
        "name": "Test Project F",
        "userIds": [
          "6001",
        ],
        "workspaceId": "1001",
      },
    ],
  },
}
`;
